core/trino-main/src/main/java/io/trino/connector/system/MaterializedViewSystemTable.java

```diff
@@ -0,0 +1,110 @@
+import io.trino.metadata.QualifiedObjectName;
+import io.trino.security.AccessControl;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorTableMetadata;
+import io.trino.spi.connector.ConnectorTransactionHandle;
+import io.trino.spi.connector.InMemoryRecordSet;
+import io.trino.spi.connector.MaterializedViewFreshness;
+import io.trino.spi.connector.RecordCursor;
+import io.trino.spi.connector.SystemTable;
+import io.trino.spi.predicate.TupleDomain;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.connector.system.jdbc.FilterUtil.tablePrefix;
+import static io.trino.connector.system.jdbc.FilterUtil.tryGetSingleVarcharValue;
+import static io.trino.metadata.MetadataListing.getMaterializedViews;
+import static io.trino.metadata.MetadataListing.listCatalogs;
+import static io.trino.spi.connector.SystemTable.Distribution.SINGLE_COORDINATOR;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewSystemTable
+        implements SystemTable
+{
+    private final Metadata metadata;
+    private final AccessControl accessControl;
+    private final MaterializedViewsSystemTableAdapter materializedViewAdapter;
+
+    @Inject
+    public MaterializedViewSystemTable(
+            Metadata metadata,
+            AccessControl accessControl,
+            MaterializedViewsSystemTableAdapter materializedViewAdapter)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.materializedViewAdapter = requireNonNull(materializedViewAdapter, "materializedViewAdapter is null");
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return materializedViewAdapter.getTableMetadata();
+    }
+
+    @Override
+    public RecordCursor cursor(
+            ConnectorTransactionHandle transactionHandle,
+            ConnectorSession connectorSession,
+            TupleDomain<Integer> constraint)
+    {
+        Session session = ((FullConnectorSession) connectorSession).getSession();
+
+        List<MaterializedViewInfo> materializedViews = getAllMaterializedViews(session, constraint);
+        List<Object[]> materializedViewRows = materializedViewAdapter.toTableRows(materializedViews);
+
+        InMemoryRecordSet.Builder displayTable = InMemoryRecordSet.builder(getTableMetadata());
+        materializedViewRows.forEach(displayTable::addRow);
+
+        return displayTable.build().cursor();
+    }
+
+    private List<MaterializedViewInfo> getAllMaterializedViews(Session session, TupleDomain<Integer> constraint)
+    {
+        Optional<String> catalogNameFilter = tryGetSingleVarcharValue(constraint, 0);
```

Comment:
The 0, 1, 2 here are based on the implicit understanding that those are the columns for catalog, schema and table. This is fine for other tables as the layout is fixed in those classes, but here the table layout comes from `materializedViewAdapter`. So now we should have `Optional<String> getCatalogNameFilter(TupleDomain<Integer> constraint)` etc. interfaces on `MaterializedViewsSystemTableAdapter` to get these filters.

Reply:
Good idea